### PR TITLE
opam tree: Allow packages with a specific version, directories or local opam files, as input

### PR DESCRIPTION
**Pull request by kit-ty-kate**
Fixes https://github.com/ocaml/opam/issues/5493
Fixes https://github.com/ocaml/opam/issues/5287

Additional questions:
* do we want to support `--recursive` and `--subpath` ?
* do we want to support `--with-version` ?

my personal opinion for those is to say *"we can merge this now, open an issue for those options and leave it for later"*

----

TODO:
* [x] Add some reftests

### CODE REVIEW

**Comment by rjbou** [comment]
I think we should support `--recursive` && `--subpath` as they are supported by other commands that permit directory lookup for opam files (lint, upgrade, install, etc.).
For `--with-version`, it is a `pin` specific option, so we shouldn't support.

These command shortcuts, ie have access to `opam install/lint .` instead of `opam pin . ; opam install/lint <pkgs>`, ease a lot the usage, but complexify set of handled option. Plus usually it is only a subset of flags that are ported, so sometimes users expect to have such option but it is not available. It is not a simple subject.

File: tests/reftests/tree.test [hunk 1]
```diff
@@ -388,3 +388,83 @@ Done.
...
+    '-- a.1 [*]
+
+local-pkg-core.dev
+### opam tree ./local-pkg-core.opam
+The following actions are simulated:
+=== install 1 package
+  - install local-pkg-core dev
+
+local-pkg-core.dev
+'-- a.1
```
Comment (rjbou) [inline tests/reftests/tree.test:470]: Some other tests
```suggestion
'-- a.1
### <pin:dir/installed-local.opam>
opam-version: "2.0"
depends: ["b" "installed-local-core"]
### <pin:dir/installed-local-core.opam>
opam-version: "2.0"
depends: "a"
### opam install ./dir
### opam tree installed-local
### opam tree ./dir
### opam tree ./dir/installed-local-core.opam
```

File: src/client/opamTreeCommand.ml [hunk 2]
```diff
@@ -430,9 +428,15 @@ let raw_state tog st names =
...
-let run st tog ?no_constraint ?(no_switch=false) mode filter names =
+let run st tog ?no_constraint ?(no_switch=false) mode filter atoms =
   let select, missing =
-    List.partition (OpamSwitchState.is_name_installed st) names
+    List.fold_left (fun (select, missing) atom ->
+        if OpamPackage.Set.disjoint
+            (OpamSwitchState.packages_of_atoms st [atom])
+            st.installed
+        then (select, atom :: missing)
+        else (fst atom :: select, missing)
```
[the remark this answers is not in the document]
Comment (kit-ty-kate) [reply src/client/opamTreeCommand.ml:438]: disjoint should be a lot faster

[the remark this answers is not in the document]
Comment (rjbou) [reply src/client/opamTreeCommand.ml:438]: `OpamSwitchState.atoms_of_package` does a check on `st.packages` while this version is applied to a much smaller set `st.installed`.

File: tests/reftests/tree.test (hunk 1 above)
Comment (kit-ty-kate) [reply tests/reftests/tree.test:470]: What is this testing? Aren't installed packages already tested above?

**Review by rjbou — APPROVED** [review]
Thanks!

**Pull request merged**